tensorflow/core/grappler/optimizers/implementation_selector.cc

```diff
@@ -261,12 +272,37 @@ Status ImplementationSelector::MaybeOptimizeFunctionCall(
   }
 
   DeviceNameUtils::ParsedName parsed_name;
-  if (!DeviceNameUtils::ParseFullName(node_def->device(), &parsed_name) ||
-      !parsed_name.has_type) {
-    return errors::Internal("Could not parse device name:", node_def->device());
+  if (!node_def->device().empty()) {
+    if (!DeviceNameUtils::ParseFullName(node_def->device(), &parsed_name) ||
+        !parsed_name.has_type) {
+      return errors::Internal("Could not parse device name:",
+                              node_def->device());
+    }
+    VLOG(2) << "Op " << node_def->name() << " runs on " << node_def->device()
+            << " = (" << parsed_name.type << ")";
   }
-  VLOG(2) << "Op " << node_def->name() << " runs on " << node_def->device()
-          << " = (" << parsed_name.type << ")";
+
+  auto select_device = [&](const string& function_name,
+                           const std::vector<string>& equiv_func_names) {
+    StringPiece device_type;
+    if (parsed_name.has_type)
+      return StringPiece(parsed_name.type);
+    else if (!cluster)
+      return StringPiece();
+    else if (const DeviceSet* device_set = cluster->GetDeviceSet()) {
+      absl::flat_hash_set<StringPiece> specified_devices;
+      specified_devices.emplace(
+          lib_info_->GetApiInfo(function_name)->preferred_device());
+      for (const string& func_name : equiv_func_names)
+        specified_devices.emplace(
+            lib_info_->GetApiInfo(func_name)->preferred_device());
+      for (const std::pair<DeviceType, int32>& dt :
+           device_set->prioritized_device_types())
+        if (specified_devices.contains(dt.first.type_string()))
+          return StringPiece(dt.first.type_string());
+    }
```

Comment:
Please put brackets `{}` around the code in each `if` branch, even if they are one-liners.

Reply:
Fixed.